syntax = "proto3";

import "google/protobuf/empty.proto";

service Seal {
    // Setup the remote server by initializing the oram blocks.
    rpc setup (SetupMessage) returns (google.protobuf.Empty) {}

    rpc search(SearchMessage) returns (SearchResponse) {}

    // Access to oram block.
    rpc oram_access(OramAccessMessage) returns (OramAccessResponse) {}

    // Init the oram block.
    rpc oram_init(OramInitMessage) returns (google.protobuf.Empty) {}
}

message SetupMessage
{
    int32 bucket_size = 1;
    int32 block_number = 2;
    int32 block_size = 3;
    int32 oram_block_size = 4;
}

message SearchMessage
{
    repeated bytes buffer = 1;
}

message SearchResponse
{
    repeated bytes buffer = 1;
}

message OramAccessMessage
{
    bool operation = 1;
    bool is_odict = 2;
    repeated bytes buffer = 3;
    int32 oram_id = 4;
}

message OramAccessResponse
{
    repeated bytes buffer = 1;
}

message OramInitMessage
{
    int32 oram_id = 1;
    int32 block_size = 2;
}

message OdictInitMessage
{
    int32 block_size = 1;
}